sljit_src/sljitExecAllocator.c

```diff
@@ -192,6 +192,63 @@ static SLJIT_INLINE void free_chunk(void *chunk, sljit_uw size)
 
 #endif
 
+#if defined(__APPLE__) && defined(__arm64__)
+__thread jit_protect_mode arm_current_jit_protect_mode = JPM_ENABLED;
+
+inline void sljit_arm_jit_write_protect_enable()
+{
+	if (__builtin_available(macOS 11, *)) {
+		if (arm_current_jit_protect_mode != JPM_ENABLED) {
+			pthread_jit_write_protect_np(1);
+			arm_current_jit_protect_mode = JPM_ENABLED;
+		}
+	}
+}
+
+inline void sljit_arm_jit_write_protect_disable()
+{
+	if (__builtin_available(macOS 11, *)) {
+		if (arm_current_jit_protect_mode != JPM_DISABLED) {
+			pthread_jit_write_protect_np(0);
+			arm_current_jit_protect_mode = JPM_DISABLED;
+		}
+	}
+}
+
+#define SLJIT_SCOPE_ENABLE_JIT_WRITE() \
+	__attribute__((unused, cleanup(sljit_arm_restore_jit_protect_mode))) \
+	jit_protect_mode scope_restrict_mode = arm_current_jit_protect_mode; \
+		sljit_arm_jit_write_protect_disable();
+
+#define SLJIT_SCOPE_ENABLE_JIT_EXEC() \
+	__attribute__((unused, cleanup(sljit_arm_restore_jit_protect_mode))) \
+	jit_protect_mode scope_restrict_mode = arm_current_jit_protect_mode; \
+		sljit_arm_jit_write_protect_enable();
+
+inline void sljit_arm_restore_jit_protect_mode(jit_protect_mode* previous_jit_protect_mode)
+{
+	if (*previous_jit_protect_mode == arm_current_jit_protect_mode)
```

Comment:
I still don't get why we need to remember the previous mode. I don't know any nested compiling in sljit.

Reply:
@zherczeg ahh right – this state is internal to sljit, so I'll remove it 👍

Reply:
However, I would point out that `sljit_malloc_exec` requires the switch from `RX` to `RW` as it writes to those pages.

Reply:
@zherczeg regarding:

> @stuartcarnie thank you for the patch. I think we don't need any automatic switching between modes. You should enable writing before `SLJIT_MALLOC_EXEC` in `sljit_generate_code` and disable it before `SLJIT_CACHE_FLUSH`. You need to do a similar thing either in `sljit_set_jump_addr` and `sljit_set_const` or in `modify_imm64_const`. Since the whole thing seems arm64 and apple specific, I would simply add a macro only to this file, define it at the beginning and undef it at the end.

Is missing some important areas:

1. `sljit_free_exec` also requires write access to the pages to add them to the free list
2. `SLJIT_MALLOC_EXEC` is used directly in the tests, meaning significant changes to those

For example, `test_exec_allocator` expects to write to memory:

https://github.com/zherczeg/sljit/blob/96d945e76e614a8bcc70d3c95acbc97e85086a89/test_src/sljitTest.c#L92-L98

Remembering the state ensures the `SLJIT_MALLOC_EXEC` does not remove the disabled write protection on exit for these tests. 

I would argue that given `SLJIT_MALLOC_EXEC` is the narrowest scope that requires RW access, it is most secure to keep `SLJIT_SCOPE_ENABLE_JIT_WRITE` macro in the `sljit_malloc_exec` function.

Reply:
These are artifical tests, and they should not define requirements for the implementation. The easiest way to handle them is disabling them on Apple systems. A better approach would be to add the necessary macro calls which are simply nops on other systems.

Reply:
@zherczeg ok, I can do that. Thanks for the feedback.

You can appreciate my concern that if I have to make a number of changes to files that look like tests, it indicates a change in behavior that may break other clients of this code. I naturally trust your confidence that these tests are not representative of a typical consumer of sljit

Reply:
`SLJIT_MALLOC_EXEC` only used in `test_exec_allocator`. Probably disabling execution at the beginning of the test and enable it at the end should be enough (that test does not execute anything). If you need modification in any other test let me know.

Reply:
not sure how it behaves in real hardware but it is likely that every new page allocated will be non-writeable until the call to the apple API and regardless if it was called before the allocation.